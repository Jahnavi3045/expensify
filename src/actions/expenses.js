import { v4 as uuidv4 } from 'uuid';
import database from '../firebase/firebase';
import { push, ref ,onValue,get, remove, update} from 'firebase/database';
import { useDispatch } from 'react-redux';

export const addExpense=(expense)=>({
    type:'ADD_EXPENSE',
    expense   
});

export const startAddExpense=(expenseData={})=>{
    // const dispatch=useDispatch();
    return(dispatch)=>{
        const {
            description='',
            note='',
            amount=0,
            createdAt=0
        }=expenseData;

        const expense={description,note,amount,createdAt}
        return push(ref(database,'expense'),expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    }
}
export const removeExpense=({id}={})=>({
    type:'REMOVE_EXPENSE',
    id
})

export const startRemoveExpense=({id}={})=>{
    return (dispatch)=>{
        return remove(ref(database,`expense/${id}`)).then(()=>{
            dispatch(removeExpense({id}))
        })
    }
}

export const editExpense=(id,updates)=>({
    type:'EDIT_EXPENSE',
    id,
    updates
})

export const startEditExpense=(id,updates)=>{
    return(dispatch)=>{
        return update(ref(database,`expense/${id}`),updates).then(()=>{
            dispatch(editExpense(id,updates));
        })
    }
}

export const setExpense=(expenses)=>({
    type:'SET_EXPENSES',
    expenses
})

export const startSetExpenses = () => {
    return async (dispatch) => {
      try {
        const expenses = [];
        const snapshot = await get(ref(database, 'expense')); // Use get() to fetch once
        
        if (snapshot.exists()) {
            snapshot.forEach((childSnapshot) => {
              expenses.push({
                id: childSnapshot.key,
                ...childSnapshot.val(),
              });
            });
            console.log(expenses)
            dispatch(setExpense(expenses));
            console.log(expenses)
        } 
      } catch (error) {
        console.error('Error fetching expenses:', error);
      }
    };
  };