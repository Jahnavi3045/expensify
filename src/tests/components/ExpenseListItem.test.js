import React from "react";
import  ExpenseListItem  from "../../components/ExpenseListItem";
import expenses from '../fixtures/expenses'
import ReactShallowRenderer from 'react-test-renderer/shallow'

test('should render expense items',()=>{
    const renderer=new ReactShallowRenderer()
    renderer.render(<ExpenseListItem {...expenses[1]}/>)
    expect(renderer.getRenderOutput()).toMatchSnapshot()
})