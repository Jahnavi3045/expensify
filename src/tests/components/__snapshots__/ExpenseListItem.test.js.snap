// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense items 1`] = `
<div>
  <Link
    to="/editExpense/2"
  >
    <h3>
      wifi bill
    </h3>
  </Link>
  <p>
    $8.00
    -
    January 1st,1970
  </p>
</div>
`;
